### PR TITLE
copypaste_based_on_bbox

### PR DESCRIPTION
**Pull request by JarvisKevin**
image

Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

It's much more expensive for us to annotate a mask label than bbox. So, How copypaste can be utilized when only bbox provided?

## Modification

If only gt_bboxes provided, gt_masks can be generated by the gt_bboxes to get a second best preformance.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
Hi @JarvisKevin 

Thanks for your kind PR. It seems that CLA is not signed. Could you sign the CLA so that eventually we could merge this PR after review? You can check the contents and follow the instruction in the communication box shown as below
image

**Comment by JarvisKevin** [comment]
[MMDet] copypaste with bbox - Jarvis的文章 - 知乎
https://zhuanlan.zhihu.com/p/568957551

**Comment by JarvisKevin** [comment]
Is there any idea to test this change in test_transforms.py?

**Comment by Czm369** [comment]
Thanks for you kind PR! CopyPaste with the PR can be directly used in object detection. Can you show images with CopyPaste based on bboxes? In addition, verify that CopyPaste can be used normally with `with_mask=Fasle` for `LoadAnnotations` in the unit test, and there is no gt_masks in the returned results.

**Comment by JarvisKevin** [comment]
@Czm369 Thanks for your suggestion of the unit_test. The picture with cp based on bboxes was posted on https://zhuanlan.zhihu.com/p/568957551, maybe you can refer it.

**Comment by Czm369** [comment]
The unit test for CopyPaste is at https://github.com/open-mmlab/mmdetection/blob/9d3e162459590eee4cfc891218dfbb5878378842/tests/test_data/test_pipelines/test_transform/test_transform.py#L1030

File: mmdet/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -2813,6 +2814,42 @@ def get_indexes(self, dataset):
...
+        xmax, ymax = bboxes[:, 2:3], bboxes[:, 3:4]
+        gt_masks = np.zeros((len(bboxes), img_h, img_w), dtype=np.uint8)
+        for i in range(len(bboxes)):
+            gt_masks[i,
+                     int(ymin[i]):int(ymax[i]),
+                     int(xmin[i]):int(xmax[i])] = 1
+        return BitmapMasks(gt_masks, img_h, img_w)
+
+    def check_gt_masks(self, results):
+        """Check gt_masks in result.
```
Comment (Czm369) [inline mmdet/datasets/pipelines/transforms.py:2838]: results

File: mmdet/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -2813,6 +2814,42 @@ def get_indexes(self, dataset):
...
+        xmin, ymin = bboxes[:, 0:1], bboxes[:, 1:2]
+        xmax, ymax = bboxes[:, 2:3], bboxes[:, 3:4]
+        gt_masks = np.zeros((len(bboxes), img_h, img_w), dtype=np.uint8)
+        for i in range(len(bboxes)):
+            gt_masks[i,
+                     int(ymin[i]):int(ymax[i]),
+                     int(xmin[i]):int(xmax[i])] = 1
+        return BitmapMasks(gt_masks, img_h, img_w)
+
+    def check_gt_masks(self, results):
```
Comment (Czm369) [inline mmdet/datasets/pipelines/transforms.py:2837]: get_gt_masks

File: mmdet/datasets/pipelines/transforms.py [hunk 3]
```diff
@@ -2813,6 +2814,42 @@ def get_indexes(self, dataset):
...
+        """Check gt_masks in results.
+
+        If gt_masks is not contained in results,
+        it will be generated based on gt_bboxes.
+        Args:
+            results (dict): Result dict.
+        Returns:
+            dict: gt_masks, originally or generated based on bboxes.
```
Comment (Czm369) [inline mmdet/datasets/pipelines/transforms.py:2845]: BitmapMasks:  gt_masks, originally or generated based on bboxes.

File: mmdet/datasets/pipelines/transforms.py [hunk 4]
```diff
@@ -2826,6 +2863,13 @@ def __call__(self, results):
         num_images = len(results['mix_results'])
         assert num_images == 1, \
             f'CopyPaste only supports processing 2 images, got {num_images}'
+
+        # check gt_masks
+        results['gt_masks'] = self.check_gt_masks(results)
```
Comment (Czm369) [inline mmdet/datasets/pipelines/transforms.py:2868]: get_gt_masks

File: mmdet/datasets/pipelines/transforms.py [hunk 5]
```diff
@@ -2813,6 +2814,42 @@ def get_indexes(self, dataset):
...
+        xmax, ymax = bboxes[:, 2:3], bboxes[:, 3:4]
+        gt_masks = np.zeros((len(bboxes), img_h, img_w), dtype=np.uint8)
+        for i in range(len(bboxes)):
+            gt_masks[i,
+                     int(ymin[i]):int(ymax[i]),
+                     int(xmin[i]):int(xmax[i])] = 1
+        return BitmapMasks(gt_masks, img_h, img_w)
+
+    def get_gt_masks(self, results):
+        """Check gt_masks in results.
```
Comment (Czm369) [inline mmdet/datasets/pipelines/transforms.py:2838]: Get gt_masks originally or generated based on bboxes.

File: mmdet/datasets/pipelines/transforms.py [hunk 6]
```diff
@@ -2813,6 +2814,42 @@ def get_indexes(self, dataset):
...
+        xmax, ymax = bboxes[:, 2:3], bboxes[:, 3:4]
+        gt_masks = np.zeros((len(bboxes), img_h, img_w), dtype=np.uint8)
+        for i in range(len(bboxes)):
+            gt_masks[i,
+                     int(ymin[i]):int(ymax[i]),
+                     int(xmin[i]):int(xmax[i])] = 1
+        return BitmapMasks(gt_masks, img_h, img_w)
+
+    def get_gt_masks(self, results):
+        """get gt_masks originally or generated based on bboxes.
```
Comment (Czm369) [inline mmdet/datasets/pipelines/transforms.py:2838]: Get gt_masks

**Review by Czm369 — APPROVED** [review]
LGTM

File: mmdet/datasets/pipelines/transforms.py [hunk 7]
```diff
@@ -2802,6 +2802,7 @@ def __init__(
         self.bbox_occluded_thr = bbox_occluded_thr
         self.mask_occluded_thr = mask_occluded_thr
         self.selected = selected
+        self.mask_gen = False
```
Comment (Czm369) [inline mmdet/datasets/pipelines/transforms.py:2805]: Add mask_gen in parameters，and claim that copypaste base on bboxes is equal to cutmix for detection in docstrings.

Comment (Czm369) [reply mmdet/datasets/pipelines/transforms.py:2805]: Add mask_gen in __repr__

Comment (Czm369) [inline mmdet/datasets/pipelines/transforms.py:2805]: Rename `mask_gen` as `using_cutmix` ?

File: mmdet/datasets/pipelines/transforms.py [hunk 8]
```diff
@@ -2789,19 +2789,22 @@ class CopyPaste:
             all objects of the source image will be pasted to the
             destination image.
             Default: True.
+        using_cutmix (bool): Whether using cutmix or not. BTW, cutmix in
```
Comment (hhaAndroid) [inline mmdet/datasets/pipelines/transforms.py:2792]: Why do we need this parameter?

Comment (JarvisKevin) [reply mmdet/datasets/pipelines/transforms.py:2792]: With the insight from @Czm369, the copypaste based on bbox is equal to the cutmix in object detection task. So, we rename the original param "mask_gen", which controls the copypaste wheather based on bbox or not, to "using_cutmix".

Comment (ZwwWayne) [reply mmdet/datasets/pipelines/transforms.py:2792]: using_cutmix is a bit vague and not clear to users. Suggest simply using `paste_by_box`, then users will know originally it is pasted by mask, and when it is on, this paste is done by box.

Comment (JarvisKevin) [reply mmdet/datasets/pipelines/transforms.py:2792]: Thanks for your suggestion. I think `paste_by_box` is a better choice.

Comment (JarvisKevin) [reply mmdet/datasets/pipelines/transforms.py:2792]: I just reviewed the logic of the code and found @hhaAndroid is right. We don't need this parameter!

The function `get_gt_masks` will help us to detect whether we have 'mask' annotations or not. If not, `gt_masks` will be automatically generated by `gen_masks_from_bboxes`.  And, It's more reasonable to `pop` the key `gt_masks` in the final `results` . So, an indicator is needed to point out whether the `gt_masks` is artificial or not. In the original version, `mask_gen` was playing such indicator role. 

In my opinion, we don't need any parameter, whatever `using_cutmix` or `paste_by_box`. @Czm369 @ZwwWayne

Comment (ZwwWayne) [reply mmdet/datasets/pipelines/transforms.py:2792]: Agree, we can do that.

**Review by hhaAndroid — APPROVED** [review]

**Comment by hhaAndroid** [comment]
@JarvisKevin Please fix the lint

File: mmdet/datasets/pipelines/transforms.py [hunk 9]
```diff
@@ -2813,6 +2814,42 @@ def get_indexes(self, dataset):
         """
         return random.randint(0, len(dataset))
 
+    def gen_masks_from_bboxes(self, bboxes, img_shape):
```
Comment (hhaAndroid) [inline mmdet/datasets/pipelines/transforms.py:2817]: Does the dev-3.x branch need to be updated in sync?

Comment (ZwwWayne) [reply mmdet/datasets/pipelines/transforms.py:2817]: Yes, we should migrate the change.

**Review by ZwwWayne — APPROVED** [review]

**Comment by ZwwWayne** [comment]
We need to migrate this PR to MMDet 3.x

**Pull request merged**